doc/dev/crimson/poseidonstore.rst

```diff
@@ -0,0 +1,352 @@
+
+
+Crash consistency
+------------------
+Large case
+- 1. Crash occurs right after writing Data blocks
+  - Data partition --> | Data blocks |
+  - We don't need to care this case. Data is not alloacted yet in reality. The blocks will be reused.
+- 2. Crash occurs right after WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - Data partition --> | Data blocks |
+  - Write procedure is completed, so there is no data loss or inconsistent state
+- 3. Crash occurs after a next write
+  - WAL --> | TxBegin A | Log Entry | TxEnd A | TxBegin B | LogEntry |
+  - Abandon failed write and recover previous state that contains Tx A
+
+Small case
+- 1. Crash occurs right after writing WAL
+  - WAL --> | TxBegin A | Log Entry| TxEnd A |
+  - All data has been written
+
+
+Comparison
+----------
+- Best case (pre-allocation)
+  - Only need two writes on both WAL and Data partition without updating Object meta and Data bitmap
+- Worst case (flush happens)
+  - Without embedding, at least three writes are required additionally on Object meta, Data bitmap, and Data blocks
+- WAL needs to be flushed if the WAL is close to full and the WAL entry is snapset, object_info_t, 
+  collection and pg_log
+  - The premise behind this design is OSD can manage the lateset metadta as a single copy. So,
+    appended entires are not to be read
+- Either best of worst case does not produce severe I/O amplification (it produce I/Os, but I/O rate is constant) 
+  unlike LSM-tree DB (proposed design is similar to LSM-tree which has only level-0)
+  
+
+With NVM
+--------
+With NVM, we are able to make use of a different scheme to store the data depending replication
+performance.
+All of the data store NVM first with out-of-place update manner, then flush them to NVMe storage device
+using in-place update scheme. This is feasible when replication latency (> 0.5 ms) is slower than
+NVMe storage performance (< 0.5 ms) because flush can store all buffered data in NVM in batch-manner.
+
+
+Detailed Design 
+===============
+
+
+- Onode lookup
+Upon onode lookup requests
+First, we hash CID (collection ID) to get the hash index of the hash table. Note that the entire hash table can be 
+pre-allocated when we limit the maximum number of objects that can be created in an AG. With the hash index, 
+the buckets are scanned until it finds the hash entry for the target object ID.
+A hash entry contains <CID, OID, Type, Offset, Segment>.
+With the hash entry, we can find the disk location of the target onode.
+Since the OID meta entry in the segment contains <OID, Chunk Number, (Offset, len), (Offset, len), … >, 
+we can read the onode from the target chunk.
+
+- Allocation
+Allocation Groups
+Entire disk space is divided into a number of equally sized chunks called Allocation Groups (AG).
+Each reactor thread in crimson-osd is responsible for a set of AGs (# of AGs / # of threads).
+Each AG has its own data structures to manage the disk partition.
```

Comment:
Are cross-AG transactions possible?  How would you plan on moving free space between reactors in the event that there are usage imbalances.  I'm fine with some vagueness here, but I'd like some design hints.

Reply:
See below